#!/bin/bash -l
#PBS -l walltime=30:00:00
#PBS -l mem=10gb
#PBS -l nodes=1:ppn=2
#PBS -M yan.zhao1@student.kuleuven.be
#PBS -m eab
#PBS -A lp_joris_vermeesch_c1

SAMPLE="HG002_sc_PTA"
PROJECT_DIR="/lustre1/project/stg_00019/research/yan/nanopore_data/2_sc_PTA";
OUTPUT_DIR="$PROJECT_DIR/6_variant_calling";
mkdir -p $OUTPUT_DIR

# Clair3 is a small variant caller for long-reads
conda activate clair3

BAM="$PROJECT_DIR/5_mapped/$SAMPLE.trimmed.sorted.bam"
REF="/lustre1/project/stg_00019/genome/homo_sapiens/hg38/genome.fa"
MODEL_PATH="/vsc-hard-mounts/leuven-data/339/vsc33900/miniconda3/envs/clair3/bin/models"

# WhatsHap is a read-based phasing tool.  (default: do not phase indels)
# WhatsHap: You can also phase indels by adding the option --indels

/vsc-hard-mounts/leuven-data/339/vsc33900/miniconda3/envs/clair3/bin/run_clair3.sh \
--python="/vsc-hard-mounts/leuven-data/339/vsc33900/miniconda3/envs/clair3/bin/python3.6" \
--parallel="/vsc-hard-mounts/leuven-data/339/vsc33900/miniconda3/envs/clair3/bin/parallel" \
--bam_fn=$BAM \
--ref_fn=$REF \
--output=$OUTPUT_DIR \
--threads=30 \
--platform="ont" \
--model_path=$MODEL_PATH/r941_prom_sup_g506 \
--sample_name=$SAMPLE \
--enable_phasing  
