#!/bin/bash -l
#PBS -l walltime=4:00:00
#PBS -l mem=10gb
#PBS -l nodes=1:ppn=2
#PBS -M yan.zhao1@student.kuleuven.be
#PBS -m eab
#PBS -A lp_joris_vermeesch_c1

# HG002_sc.fastq.gz 52G

module load Python/3.7.2-foss-2018a  

PROJECT_DIR="/lustre1/project/stg_00019/research/yan/nanopore_data/sc_repli-g";
SAMPLE_DIR="$PROJECT_DIR/1_merged_fastq";
OUTPUT_DIR="$PROJECT_DIR/3_nanofilt_trimmed";
SAMPLE="HG002_sc_PTA"

mkdir -p $OUTPUT_DIR;
cd $OUTPUT_DIR;
rsync -ahrL $SAMPLE_DIR/$SAMPLE.fastq.gz .;
gunzip *gz;

# remove all sequences shorter than 500 nucleotides (option -l)
# trim the first 9 nucleotides off all reads (option â€“headcrop)
# SNP identification: you don't want too many low quality bases

/vsc-hard-mounts/leuven-data/339/vsc33900/miniconda3/bin/NanoFilt -q 9 -l 500 --headcrop 9 < $SAMPLE.fastq|gzip > $SAMPLE.trimmed.fastq.gz