#!/bin/bash -l
#PBS -l walltime=10:00:00
#PBS -l mem=30gb
#PBS -l nodes=1:ppn=10
#PBS -M yan.zhao1@student.kuleuven.be
#PBS -m eab
#PBS -A lp_joris_vermeesch_c1

PROJECT_DIR="/lustre1/project/stg_00019/research/yan/nanopore_data/sc_repli-g";
SAMPLE_DIR="$PROJECT_DIR/3_nanofilt_trimmed";
OUTPUT_DIR="$PROJECT_DIR/5_mapped";
REF="/lustre1/project/stg_00019/genome/homo_sapiens/hg38/genome.fa"
SAMPLE="HG002_sc"

# minimap2 can mapping PacBio or Oxford Nanopore genomic reads to the human genome;
module load minimap2/2.12-foss-2018a
module load SAMtools/1.9-GCC-6.4.0-2.28

mkdir -p $OUTPUT_DIR

cd $OUTPUT_DIR

#  -a output in the SAM format
#  -R STR SAM read group line in a format like '@RG\tID:foo\tSM:bar' 
# -R "@RG\tID:$SAMPLE\tSM:$SAMPLE"

minimap2 -a -R "@RG\tID:$SAMPLE\tSM:$SAMPLE" -t 20 $REF $SAMPLE_DIR/$SAMPLE.trimmed.fastq.gz > $SAMPLE.trimmed.sam

# Convert to bam file
samtools view -Shb $SAMPLE.trimmed.sam > $SAMPLE.trimmed.bam
# rm $SAMPLE.trimmed.sam 

# Sort the bam file
samtools sort -o $SAMPLE.trimmed.sorted.bam $SAMPLE.trimmed.bam
rm $SAMPLE.trimmed.bam

# create an index file
samtools index $SAMPLE.trimmed.sorted.bam

